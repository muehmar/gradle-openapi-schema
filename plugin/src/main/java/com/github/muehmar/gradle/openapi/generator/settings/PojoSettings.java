package com.github.muehmar.gradle.openapi.generator.settings;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.task.TaskIdentifier;
import com.github.muehmar.gradle.openapi.warnings.Warning;
import com.github.muehmar.gradle.openapi.warnings.WarningsContext;
import io.github.muehmar.pojobuilder.annotations.FieldBuilder;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.io.Serializable;
import java.util.List;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class PojoSettings implements Serializable {
  JsonSupport jsonSupport;
  String packageName;
  String suffix;
  StagedBuilderSettings stagedBuilder;
  String builderMethodPrefix;
  boolean enableValidation;
  boolean nonStrictOneOfValidation;

  ValidationApi validationApi;
  List<ClassTypeMapping> classTypeMappings;
  List<FormatTypeMapping> formatTypeMappings;
  EnumDescriptionSettings enumDescriptionSettings;
  GetterSuffixes getterSuffixes;
  ValidationMethods validationMethods;
  List<String> excludeSchemas;
  PojoNameMappings pojoNameMappings;

  TaskIdentifier taskIdentifier;

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableValidation() {
    return enableValidation;
  }

  public boolean isNonStrictOneOfValidation() {
    return nonStrictOneOfValidation;
  }

  public List<ClassTypeMapping> getClassTypeMappingsList() {
    return classTypeMappings;
  }

  public List<FormatTypeMapping> getFormatTypeMappingsList() {
    return formatTypeMappings;
  }

  public List<String> getExcludeSchemasList() {
    return excludeSchemas;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public ExcludedSchemas getExcludedSchemas() {
    return ExcludedSchemas.fromExcludedPojoNames(
        PList.fromIter(excludeSchemas).map(Name::ofString));
  }

  public boolean isEnableStagedBuilder() {
    return stagedBuilder.isEnabled();
  }

  public boolean isDisableStagedBuilder() {
    return !isEnableStagedBuilder();
  }

  public StagedBuilderSettings getStagedBuilder() {
    return stagedBuilder;
  }

  @FieldBuilder(fieldName = "classTypeMappings")
  public static List<ClassTypeMapping> classTypeMappings(
      PList<ClassTypeMapping> classTypeMappings) {
    return classTypeMappings.toArrayList();
  }

  @FieldBuilder(fieldName = "formatTypeMappings")
  public static List<FormatTypeMapping> formatTypeMappings(
      PList<FormatTypeMapping> formatTypeMappings) {
    return formatTypeMappings.toArrayList();
  }

  public TypeMappings getTypeMappings() {
    return new TypeMappings(PList.fromIter(classTypeMappings), PList.fromIter(formatTypeMappings));
  }

  public PojoNameMapping pojoNameMapping() {
    return PList.fromIter(pojoNameMappings.getConstantNameMappings())
        .map(
            constantNameMapping ->
                PojoNameMapping.replaceConstant(
                    constantNameMapping.getConstant(), constantNameMapping.getReplacement()))
        .foldLeft(PojoNameMapping.noMapping(), PojoNameMapping::andThen);
  }

  public void validate() {
    classTypeMappings.forEach(
        classTypeMapping -> {
          if (not(classTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(classTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });

    formatTypeMappings.forEach(
        formatTypeMapping -> {
          if (not(formatTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(formatTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });
  }
}
