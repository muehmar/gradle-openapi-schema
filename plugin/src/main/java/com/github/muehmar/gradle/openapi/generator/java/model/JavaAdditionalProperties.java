package com.github.muehmar.gradle.openapi.generator.java.model;

import static com.github.muehmar.gradle.openapi.generator.java.model.type.JavaAnyType.javaAnyType;
import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.AdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.Value;

@Value
public class JavaAdditionalProperties {
  private static final JavaName MAP_PROPERTY_NAME = JavaName.fromString("additionalProperties");
  boolean allowed;
  JavaType type;

  public static JavaAdditionalProperties wrap(
      AdditionalProperties additionalProperties, TypeMappings typeMappings) {
    return new JavaAdditionalProperties(
        additionalProperties.isAllowed(),
        JavaType.wrap(additionalProperties.getType(), typeMappings));
  }

  public static JavaAdditionalProperties anyTypeAllowed() {
    return new JavaAdditionalProperties(true, javaAnyType());
  }

  public static JavaAdditionalProperties allowedFor(JavaType type) {
    return new JavaAdditionalProperties(true, type);
  }

  public static JavaAdditionalProperties notAllowed() {
    return new JavaAdditionalProperties(false, javaAnyType());
  }

  public static JavaName additionalPropertiesName() {
    return MAP_PROPERTY_NAME;
  }

  public boolean isNotAllowed() {
    return not(allowed);
  }

  public boolean isValueAnyType() {
    return type.equals(javaAnyType());
  }

  public boolean isNotValueAnyType() {
    return not(type.equals(javaAnyType()));
  }

  public TechnicalPojoMember asTechnicalPojoMember() {
    return TechnicalPojoMember.additionalProperties(type);
  }

  public Optional<EnumGenerator.EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumGenerator.EnumContent>> createEnumContent =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaName.fromName(type.getQualifiedClassName().getClassName()))
                    .description("Additional property enum")
                    .members(enumType.getMembers())
                    .build());
    return type.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        createEnumContent,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }
}
