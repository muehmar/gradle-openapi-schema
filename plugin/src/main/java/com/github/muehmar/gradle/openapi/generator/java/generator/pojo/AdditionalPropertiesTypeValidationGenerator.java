package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.ValidationGenerator.assertTrue;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;
import static io.github.muehmar.codegenerator.Generator.constant;

import com.github.muehmar.gradle.openapi.generator.java.generator.shared.SettingsFunctions;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;

public class AdditionalPropertiesTypeValidationGenerator {
  private AdditionalPropertiesTypeValidationGenerator() {}

  public static Generator<JavaObjectPojo, PojoSettings>
      additionalPropertiesTypeValidationGenerator() {
    return annotatedCorrectTypeMethod()
        .filter(JavaAdditionalProperties::isAllowed)
        .filter(JavaAdditionalProperties::isNotValueAnyType)
        .contraMap(JavaObjectPojo::getAdditionalProperties);
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> annotatedCorrectTypeMethod() {
    return JacksonAnnotationGenerator.<JavaAdditionalProperties>jsonIgnore()
        .append(
            assertTrue(
                props ->
                    String.format(
                        "Not all additional properties are instances of %s",
                        props.getType().getQualifiedClassName().getClassName())))
        .append(correctTypeMethod());
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> correctTypeMethod() {
    return MethodGenBuilder.<JavaAdditionalProperties, PojoSettings>create()
        .modifiers(SettingsFunctions::validationMethodModifiers)
        .noGenericTypes()
        .returnType("boolean")
        .methodName("isAllAdditionalPropertiesHaveCorrectType")
        .noArguments()
        .content(
            constant(
                String.format(
                    "return getAdditionalProperties().size() == %s.size();",
                    additionalPropertiesName())))
        .build();
  }
}
