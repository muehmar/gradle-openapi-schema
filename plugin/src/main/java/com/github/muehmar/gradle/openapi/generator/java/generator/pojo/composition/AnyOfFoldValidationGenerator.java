package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.composition;

import static com.github.muehmar.gradle.openapi.generator.java.model.name.MethodNames.Composition.AnyOf.foldAnyOfMethodName;
import static com.github.muehmar.gradle.openapi.generator.java.model.name.MethodNames.Composition.AnyOf.getAnyOfMethodName;
import static com.github.muehmar.gradle.openapi.generator.java.model.name.MethodNames.Composition.AnyOf.getAnyOfValidCountMethodName;
import static io.github.muehmar.codegenerator.Generator.constant;

import ch.bluecare.commons.data.NonEmptyList;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.DeprecatedMethodGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.SettingsFunctions;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.validation.ValidationAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.MethodGen;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;

public class AnyOfFoldValidationGenerator {

  private AnyOfFoldValidationGenerator() {}

  public static Generator<JavaObjectPojo, PojoSettings> anyOfFoldValidationGenerator() {
    return Generator.<JavaObjectPojo, PojoSettings>emptyGen()
        .appendOptional(
            annotatedMethod(), p -> p.getAnyOfComposition().map(JavaAnyOfComposition::getPojos));
  }

  private static Generator<NonEmptyList<JavaObjectPojo>, PojoSettings> annotatedMethod() {
    final MethodGen<NonEmptyList<JavaObjectPojo>, PojoSettings> method =
        MethodGenBuilder.<NonEmptyList<JavaObjectPojo>, PojoSettings>create()
            .modifiers(SettingsFunctions::validationMethodModifiers)
            .noGenericTypes()
            .returnType("List<Object>")
            .methodName(getAnyOfMethodName().asString())
            .noArguments()
            .doesNotThrow()
            .content(methodContent())
            .build();

    return DeprecatedMethodGenerator
        .<NonEmptyList<JavaObjectPojo>>deprecatedJavaDocAndAnnotationForValidationMethod()
        .append(ValidationAnnotationGenerator.validAnnotation())
        .append(JacksonAnnotationGenerator.jsonIgnore())
        .append(method);
  }

  private static Generator<NonEmptyList<JavaObjectPojo>, PojoSettings> methodContent() {
    return Generator.<NonEmptyList<JavaObjectPojo>, PojoSettings>emptyGen()
        .append(w -> w.println("if (%s() == 0) {", getAnyOfValidCountMethodName()))
        .append(constant("return null;"), 1)
        .append(constant("}"))
        .append(
            (pojos, s, w) ->
                w.println(
                    "return %s(%s);",
                    foldAnyOfMethodName(),
                    pojos.map(name -> "dto -> dto").toPList().mkString(", ")));
  }
}
