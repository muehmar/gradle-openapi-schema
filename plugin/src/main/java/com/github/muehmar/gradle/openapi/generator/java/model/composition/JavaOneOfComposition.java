package com.github.muehmar.gradle.openapi.generator.java.model.composition;

import static com.github.muehmar.gradle.openapi.generator.java.model.composition.Assertion.assertAllObjectPojos;

import ch.bluecare.commons.data.NonEmptyList;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.TechnicalPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojo;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.composition.OneOfComposition;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaOneOfComposition {
  private final NonEmptyList<JavaObjectPojo> pojos;
  private final Optional<JavaDiscriminator> discriminator;

  JavaOneOfComposition(NonEmptyList<JavaPojo> pojos, Optional<JavaDiscriminator> discriminator) {
    this.pojos = assertAllObjectPojos(pojos);
    this.discriminator = discriminator;
  }

  public static JavaOneOfComposition wrap(
      OneOfComposition oneOfComposition,
      Optional<Discriminator> objectPojoDiscriminator,
      PojoType type,
      TypeMappings typeMappings) {
    final Optional<JavaDiscriminator> javaDiscriminator =
        oneOfComposition
            .determineDiscriminator(objectPojoDiscriminator)
            .map(JavaDiscriminator::wrap);
    return new JavaOneOfComposition(
        oneOfComposition
            .getPojos()
            .map(pojo -> JavaPojo.wrap(pojo, typeMappings))
            .map(result -> result.getTypeOrDefault(type)),
        javaDiscriminator);
  }

  public static JavaOneOfComposition fromPojos(NonEmptyList<JavaPojo> pojos) {
    return new JavaOneOfComposition(pojos, Optional.empty());
  }

  public NonEmptyList<JavaObjectPojo> getPojos() {
    return pojos;
  }

  public PList<JavaPojoMember> getMembers() {
    return pojos
        .toPList()
        .flatMap(JavaObjectPojo::getAllMembersForComposition)
        .map(JavaPojoMember::asOneOfMember)
        .distinct(Function.identity());
  }

  public PList<TechnicalPojoMember> getPojosAsTechnicalMembers() {
    return pojos.toPList().map(TechnicalPojoMember::wrapJavaObjectPojo);
  }

  public Optional<JavaDiscriminator> getDiscriminator() {
    return discriminator;
  }
}
