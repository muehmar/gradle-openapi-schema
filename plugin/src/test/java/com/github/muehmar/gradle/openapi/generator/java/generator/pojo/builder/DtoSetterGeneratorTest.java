package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.builder;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.builder.DtoSetterGenerator.dtoSetterGenerator;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo1;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo2;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.withAdditionalProperties;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.NonEmptyList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfCompositions;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixesBuilder;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

@SnapshotTest
class DtoSetterGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("oneOfPojo")
  void generator_when_calledWithComposedPojo_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = dtoSetterGenerator();

    final GetterSuffixes getterSuffixes =
        GetterSuffixesBuilder.create()
            .requiredSuffix("")
            .requiredNullableSuffix("Opt1")
            .optionalSuffix("Opt2")
            .optionalNullableSuffix("Tristate")
            .build();
    final Writer writer =
        generator.generate(
            JavaPojos.oneOfPojo(
                sampleObjectPojo1(), JavaPojos.allNecessityAndNullabilityVariants()),
            defaultTestSettings().withGetterSuffixes(getterSuffixes),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("oneOfPojoWithDiscriminator")
  void generator_when_calledWithOneOfPojoWithDiscriminator_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = dtoSetterGenerator();

    final Discriminator discriminator =
        Discriminator.fromPropertyName(
            JavaPojoMembers.requiredString().getName().getOriginalName());
    final JavaOneOfComposition javaOneOfComposition =
        JavaOneOfCompositions.fromPojosAndDiscriminator(
            NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()), discriminator);

    final Writer writer =
        generator.generate(
            JavaPojos.oneOfPojo(javaOneOfComposition), defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("noBuilderSetMethodPrefix")
  void generator_when_noBuilderSetMethodPrefix_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = dtoSetterGenerator();

    final Writer writer =
        generator.generate(
            JavaPojos.oneOfPojo(sampleObjectPojo1(), sampleObjectPojo2()),
            defaultTestSettings().withBuilderMethodPrefix(""),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("composedPojoHasNoAdditionalPropertiesAllowed")
  void generator_when_composedPojoHasNoAdditionalPropertiesAllowed_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator = dtoSetterGenerator();

    final JavaObjectPojo samplePojo1 =
        withAdditionalProperties(sampleObjectPojo1(), JavaAdditionalProperties.notAllowed());

    final Writer writer =
        generator.generate(
            JavaPojos.oneOfPojo(samplePojo1, sampleObjectPojo2()),
            defaultTestSettings().withBuilderMethodPrefix(""),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
