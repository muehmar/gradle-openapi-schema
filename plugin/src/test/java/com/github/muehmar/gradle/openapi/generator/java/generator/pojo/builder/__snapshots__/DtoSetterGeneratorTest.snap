allOfPojo=[
.
.
private Builder setSampleObjectPojo1Dto(SampleObjectPojo1Dto dto) {
  setStringVal(dto.getStringVal());
  setIntVal(dto.getIntVal());
  setDoubleVal(dto.getDoubleVal());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}

private Builder setNecessityAndNullabilityDto(NecessityAndNullabilityDto dto) {
  setRequiredStringVal(dto.getRequiredStringVal());
  setRequiredNullableStringVal(dto.getRequiredNullableStringVal());
  setOptionalStringVal(dto.getOptionalStringVal());
  setOptionalNullableStringVal(dto.getOptionalNullableStringVal());
  setRequiredListWithNullableItems_(dto.getRequiredListWithNullableItems());
  setRequiredNullableListWithNullableItems_(dto.getRequiredNullableListWithNullableItems());
  setOptionalListWithNullableItems_(dto.getOptionalListWithNullableItems());
  setOptionalNullableListWithNullableItems_(dto.getOptionalNullableListWithNullableItems());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]


anyOfPojo=[
.
.
private Builder setSampleObjectPojo1Dto(SampleObjectPojo1Dto dto) {
  setStringVal(dto.getStringVal());
  setIntVal(dto.getIntVal());
  setDoubleVal(dto.getDoubleVal());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}

private Builder setNecessityAndNullabilityDto(NecessityAndNullabilityDto dto) {
  setRequiredStringVal(dto.getRequiredStringVal());
  setRequiredNullableStringVal(dto.getRequiredNullableStringVal());
  setOptionalStringVal(dto.getOptionalStringVal());
  setOptionalNullableStringVal(dto.getOptionalNullableStringVal());
  setRequiredListWithNullableItems_(dto.getRequiredListWithNullableItems());
  setRequiredNullableListWithNullableItems_(dto.getRequiredNullableListWithNullableItems());
  setOptionalListWithNullableItems_(dto.getOptionalListWithNullableItems());
  setOptionalNullableListWithNullableItems_(dto.getOptionalNullableListWithNullableItems());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]


anyOfPojoWithDiscriminator=[
.
.
private Builder setSampleObjectPojo1Dto(SampleObjectPojo1Dto dto) {
  setStringVal("SampleObjectPojo1");
  setIntVal(dto.getIntVal());
  setDoubleVal(dto.getDoubleVal());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}

private Builder setSampleObjectPojo2Dto(SampleObjectPojo2Dto dto) {
  setStringVal("SampleObjectPojo2");
  setBirthdate(dto.getBirthdate());
  setEmail(dto.getEmail());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]


composedPojoHasNoAdditionalPropertiesAllowed=[
.
.
private Builder sampleObjectPojo1Dto(SampleObjectPojo1Dto dto) {
  stringVal(dto.getStringVal());
  intVal(dto.getIntVal());
  doubleVal(dto.getDoubleVal());
  return this;
}

private Builder sampleObjectPojo2Dto(SampleObjectPojo2Dto dto) {
  stringVal(dto.getStringVal());
  birthdate(dto.getBirthdate());
  email(dto.getEmail());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]


composedPojoHasNotNullableAdditionalProperties=[
.
.
private Builder sampleObjectPojo1Dto(SampleObjectPojo1Dto dto) {
  stringVal(dto.getStringVal());
  intVal(dto.getIntVal());
  doubleVal(dto.getDoubleVal());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue()));
  return this;
}

private Builder sampleObjectPojo2Dto(SampleObjectPojo2Dto dto) {
  stringVal(dto.getStringVal());
  birthdate(dto.getBirthdate());
  email(dto.getEmail());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]


nestedOneOfPojo=[
.
.
private Builder setSampleObjectPojo1Dto(SampleObjectPojo1Dto dto) {
  setStringVal(dto.getStringVal());
  setIntVal(dto.getIntVal());
  setDoubleVal(dto.getDoubleVal());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}

private Builder setOneOfPojo1Dto(OneOfPojo1Dto dto) {
  setRequiredStringVal(dto.getRequiredStringVal());
  if (dto.getIsRequiredNullableStringValPresent()) setRequiredNullableStringVal(dto.getRequiredNullableStringVal());
  if (dto.getIsOptionalStringValNotNull()) setOptionalStringVal(dto.getOptionalStringVal());
  setOptionalNullableStringVal(dto.getOptionalNullableStringVal());
  setRequiredListWithNullableItems_(dto.getRequiredListWithNullableItems());
  if (dto.getIsRequiredNullableListWithNullableItemsPresent()) setRequiredNullableListWithNullableItems_(dto.getRequiredNullableListWithNullableItems());
  if (dto.getIsOptionalListWithNullableItemsNotNull()) setOptionalListWithNullableItems_(dto.getOptionalListWithNullableItems());
  setOptionalNullableListWithNullableItems_(dto.getOptionalNullableListWithNullableItems());
  setStringVal(dto.getStringVal());
  setBirthdate(dto.getBirthdate());
  setEmail(dto.getEmail());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]


noBuilderSetMethodPrefix=[
.
.
private Builder sampleObjectPojo1Dto(SampleObjectPojo1Dto dto) {
  stringVal(dto.getStringVal());
  intVal(dto.getIntVal());
  doubleVal(dto.getDoubleVal());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}

private Builder sampleObjectPojo2Dto(SampleObjectPojo2Dto dto) {
  stringVal(dto.getStringVal());
  birthdate(dto.getBirthdate());
  email(dto.getEmail());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]


oneOfPojo=[
.
.
private Builder setSampleObjectPojo1Dto(SampleObjectPojo1Dto dto) {
  setStringVal(dto.getStringVal());
  setIntVal(dto.getIntVal());
  setDoubleVal(dto.getDoubleVal());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}

private Builder setNecessityAndNullabilityDto(NecessityAndNullabilityDto dto) {
  setRequiredStringVal(dto.getRequiredStringVal());
  setRequiredNullableStringVal(dto.getRequiredNullableStringValOpt1());
  setOptionalStringVal(dto.getOptionalStringValOpt2());
  setOptionalNullableStringVal(dto.getOptionalNullableStringValTristate());
  setRequiredListWithNullableItems_(dto.getRequiredListWithNullableItems());
  setRequiredNullableListWithNullableItems_(dto.getRequiredNullableListWithNullableItemsOpt1());
  setOptionalListWithNullableItems_(dto.getOptionalListWithNullableItemsOpt2());
  setOptionalNullableListWithNullableItems_(dto.getOptionalNullableListWithNullableItemsTristate());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]


oneOfPojoWithDiscriminator=[
.
.
private Builder setSampleObjectPojo1Dto(SampleObjectPojo1Dto dto) {
  setStringVal("SampleObjectPojo1");
  setIntVal(dto.getIntVal());
  setDoubleVal(dto.getDoubleVal());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}

private Builder setSampleObjectPojo2Dto(SampleObjectPojo2Dto dto) {
  setStringVal("SampleObjectPojo2");
  setBirthdate(dto.getBirthdate());
  setEmail(dto.getEmail());
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]


oneOfPojoWithEnumDiscriminator=[
.
.
private Builder setYellowDto(YellowDto dto) {
  setStringVal(dto.getStringVal());
  setColor(BaseDto.Color.YELLOW);
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}

private Builder setOrangeDto(OrangeDto dto) {
  setBirthdate(dto.getBirthdate());
  setColor(BaseDto.Color.ORANGE);
  dto.getAdditionalProperties()
      .forEach(prop -> addAdditionalProperty(prop.getName(), prop.getValue().orElse(null)));
  return this;
}
]